compositor_pipeline/src/queue.rs

```diff
@@ -0,0 +1,97 @@
+mod internal_queue;
+
+use std::{
+    sync::{Arc, Mutex},
+    thread::{self, sleep},
+    time::{Duration, Instant},
+};
+
+use compositor_common::{Frame, Framerate, InputId};
+use compositor_render::input_frames::InputFrames;
+use crossbeam_channel::{tick, unbounded, Receiver, Sender};
+
+use self::internal_queue::{InternalQueue, QueueError};
+
+const DEFAULT_BUFFER_DURATION: Duration = Duration::from_millis(10);
+
+pub struct Queue {
+    internal_queue: Arc<Mutex<InternalQueue>>,
+    check_queue_channel: (Sender<()>, Receiver<()>),
+    output_framerate: Framerate,
+    buffer_duration: Duration,
+}
+
+impl Queue {
+    pub fn new(output_framerate: Framerate) -> Self {
+        Queue {
+            internal_queue: Arc::new(Mutex::new(InternalQueue::new(output_framerate))),
+            check_queue_channel: unbounded(),
+            output_framerate,
+            buffer_duration: DEFAULT_BUFFER_DURATION,
+        }
+    }
+
+    pub fn add_input(&self, input_id: InputId) {
+        let mut internal_queue = self.internal_queue.lock().unwrap();
+        internal_queue.add_input(input_id);
+    }
+
+    #[allow(dead_code)]
+    pub fn remove_input(&self, input_id: InputId) {
+        let mut internal_queue = self.internal_queue.lock().unwrap();
+        // TODO: gracefully remove input - wait until last enqueued frame PTS is smaller than output PTS
+        internal_queue.remove_input(input_id);
+    }
+
+    #[allow(dead_code)]
+    pub fn start(&self, sender: Sender<InputFrames>) {
+        let (check_queue_sender, check_queue_receiver) = self.check_queue_channel.clone();
+        let internal_queue = self.internal_queue.clone();
+        let tick_duration = self.output_framerate.get_interval_duration();
+        let buffer_duration = self.buffer_duration;
+
+        thread::spawn(move || {
+            // Wait for first frame
+            check_queue_receiver.recv().unwrap();
+            sleep(buffer_duration);
+            Self::start_ticker(tick_duration, check_queue_sender);
+
+            let start = Instant::now();
+            loop {
```

Comment:
this is busy loop, you missed reading from channel. This would explain that 100% cpu usage you posted on slack :D

Reply:
You are right. I totally forgot about that. It's great that you catch it 😃 I will run those benchmarks once again. Now that 100% CPU usage makes sense 😛

Reply:
Well, even without this loop constantly running, encoding Big Buck Bunny in 4k 30fps was "a little bit" too much for M1 pro 😄
<img width="1512" alt="Screenshot 2023-07-24 at 21 51 33" src="https://github.com/membraneframework/video_compositor/assets/104033489/86f9540e-553b-4652-bfe6-b59478d343bd">